From cd92f2088377bd2b5145ec56f4dab50adcf29881 Mon Sep 17 00:00:00 2001
From: Debarshi Ray <debarshir@gnome.org>
Date: Mon, 12 May 2014 14:57:18 +0200
Subject: [PATCH] Restore transparency

The transparency settings were removed as a side effect of
2bff4b63ed3ceef6055e35563e9b0b33ad57349d

This restores them and you will need a compositing window manager to
use it. The background image setting, also known as faux transparency,
was not restored.

Also contains

326c4f143511a3fae61aed1466568260a2ac1c4e Mon Sep 17 00:00:00 2001
Lars Uebernickel <lars.uebernickel@canonical.com>
Wed, 28 May 2014 14:11:02 +0200
[PATCH] window: Make the drawing robust across all themes

There are lots of themes out there in the wild that do not specify a
background-color for all widgets and the default is transparent. This
is usually not a problem because GTK+ sets an opaque region on the
whole window and things without a background-color get drawn with the
theme's default background colour. However, to achieve transparency
we disable the opaque region by making the window app-paintable. This
can lead to transparent menubars or notebook tabs in some themes. We
can avoid this by ensuring that the window always renders a background.

https://bugzilla.gnome.org/show_bug.cgi?id=730016

---
 src/org.gnome.Terminal.gschema.xml |  9 ++++
 src/profile-editor.c               | 10 +++++
 src/profile-preferences.ui         | 92 ++++++++++++++++++++++++++++++++++++++
 src/terminal-schemas.h             |  3 ++
 src/terminal-screen.c              | 23 +++++++++-
 src/terminal-window.c              |  7 +++
 6 files changed, 143 insertions(+), 1 deletion(-)

Index: b/src/org.gnome.Terminal.gschema.xml
===================================================================
--- a/src/org.gnome.Terminal.gschema.xml
+++ b/src/org.gnome.Terminal.gschema.xml
@@ -370,6 +370,20 @@
       <default>'narrow'</default>
       <summary>Whether ambiguous-width characters are narrow or wide when using UTF-8 encoding</summary>
     </key>
+    <key name="use-transparent-background" type="b">
+      <default>false</default>
+      <summary>Whether to use a transparent background</summary>
+    </key>
+    <key name="use-theme-transparency" type="b">
+      <default>true</default>
+      <summary>Whether to use the value of TerminalScreen-background-darkness,
+          if available, from the theme for the transparency value.</summary>
+    </key>
+    <key name="background-transparency-percent" type="i">
+      <default>50</default>
+      <summary>Adjust the amount of transparency</summary>
+      <description>A value between 0 and 100, where 0 is opaque and 100 is fully transparent.</description>
+    </key>
   </schema>
 
   <!-- Keybinding settings -->
Index: b/src/profile-editor.c
===================================================================
--- a/src/profile-editor.c
+++ b/src/profile-editor.c
@@ -35,6 +35,7 @@
 #include "terminal-util.h"
 #include "terminal-profiles-list.h"
 #include "terminal-libgsystem.h"
+#include "terminal-window.h"
 
 typedef struct _TerminalColorScheme TerminalColorScheme;
 
@@ -822,6 +823,7 @@
   GtkWidget *editor, *w;
   gs_free char *uuid = NULL;
   guint i;
+  gfloat style_darkness;
 
   editor = g_object_get_data (G_OBJECT (profile), "editor-window");
   if (editor)
@@ -829,7 +831,7 @@
       terminal_util_dialog_focus_widget (editor, widget_name);
 
       gtk_window_set_transient_for (GTK_WINDOW (editor),
-                                    GTK_WINDOW (transient_parent));
+                                    NULL);
       gtk_window_present (GTK_WINDOW (editor));
       return;
     }
@@ -908,6 +910,15 @@
                         profile);
     }
 
+  gtk_widget_style_get (GTK_WIDGET (
+              terminal_window_get_active (TERMINAL_WINDOW (transient_parent))),
+                        "background-darkness", &style_darkness,
+                        NULL);
+
+  gtk_widget_set_visible (gtk_builder_get_object (
+              builder,
+              "use-theme-transparency-checkbutton"), style_darkness >= 0);
+
   profile_palette_notify_colorpickers_cb (profile, TERMINAL_PROFILE_PALETTE_KEY, editor);
   g_signal_connect (profile, "changed::" TERMINAL_PROFILE_PALETTE_KEY,
                     G_CALLBACK (profile_palette_notify_colorpickers_cb),
@@ -1218,12 +1229,33 @@
                    "active-id",
                    G_SETTINGS_BIND_GET | G_SETTINGS_BIND_SET);
 
+  g_settings_bind (profile, TERMINAL_PROFILE_USE_TRANSPARENT_BACKGROUND,
+                   gtk_builder_get_object (builder, "use-transparent-background"),
+                   "active", G_SETTINGS_BIND_GET | G_SETTINGS_BIND_SET);
+  g_settings_bind (profile, TERMINAL_PROFILE_USE_TRANSPARENT_BACKGROUND,
+                   gtk_builder_get_object (builder, "background-transparent-scale-box"),
+                   "sensitive", G_SETTINGS_BIND_GET | G_SETTINGS_BIND_NO_SENSITIVITY);
+  g_settings_bind (profile, TERMINAL_PROFILE_BACKGROUND_TRANSPARENCY_PERCENT,
+                   gtk_builder_get_object (builder, "background-transparent-adjustment"),
+                   "value", G_SETTINGS_BIND_GET | G_SETTINGS_BIND_SET);
+  g_settings_bind (profile, TERMINAL_PROFILE_USE_THEME_TRANSPARENCY,
+                   gtk_builder_get_object (builder, "use-theme-transparency-checkbutton"),
+                   "active", G_SETTINGS_BIND_GET | G_SETTINGS_BIND_SET);
+
+  if (style_darkness >= 0)
+      g_settings_bind (profile, TERMINAL_PROFILE_USE_THEME_TRANSPARENCY,
+              gtk_builder_get_object (builder, "use-transparent-background-box"),
+              "sensitive",
+              G_SETTINGS_BIND_GET |
+              G_SETTINGS_BIND_INVERT_BOOLEAN |
+              G_SETTINGS_BIND_NO_SENSITIVITY);
+
   /* Finished! */
   terminal_util_bind_mnemonic_label_sensitivity (editor);
 
   terminal_util_dialog_focus_widget (editor, widget_name);
 
   gtk_window_set_transient_for (GTK_WINDOW (editor),
-                                GTK_WINDOW (transient_parent));
+                                NULL);
   gtk_window_present (GTK_WINDOW (editor));
 }
Index: b/src/profile-preferences.ui
===================================================================
--- a/src/profile-preferences.ui
+++ b/src/profile-preferences.ui
@@ -41,6 +41,11 @@
       </row>
     </data>
   </object>
+  <object class="GtkAdjustment" id="background-transparent-adjustment">
+    <property name="upper">100</property>
+    <property name="step_increment">1</property>
+    <property name="page_increment">10</property>
+  </object>
   <object class="GtkListStore" id="model1">
     <columns>
       <!-- column-name gchararray -->
@@ -1035,6 +1040,122 @@
                                 <property name="position">1</property>
                               </packing>
                             </child>
+                            <child>
+                              <object class="GtkBox" id="use-transparent-background-box">
+                                <property name="visible">True</property>
+                                <property name="can_focus">False</property>
+                                <property name="orientation">horizontal</property>
+                                <property name="spacing">12</property>
+                                <child>
+                                  <object class="GtkCheckButton" id="use-transparent-background">
+                                    <property name="label" translatable="yes">Use t_ransparent background</property>
+                                    <property name="visible">True</property>
+                                    <property name="can_focus">True</property>
+                                    <property name="receives_default">False</property>
+                                    <property name="use_underline">True</property>
+                                    <property name="xalign">0</property>
+                                    <property name="draw_indicator">True</property>
+                                  </object>
+                                  <packing>
+                                    <property name="expand">False</property>
+                                    <property name="fill">False</property>
+                                    <property name="position">0</property>
+                                  </packing>
+                                </child>
+                                <child>
+                                  <object class="GtkBox" id="background-transparent-scale-box">
+                                    <property name="visible">True</property>
+                                    <property name="can_focus">False</property>
+                                    <property name="orientation">horizontal</property>
+                                    <property name="spacing">6</property>
+                                    <child>
+                                      <object class="GtkLabel" id="background-transparent-min-label">
+                                        <property name="visible">True</property>
+                                        <property name="can_focus">False</property>
+                                        <property name="xalign">0.5</property>
+                                        <property name="label" translatable="yes">none</property>
+                                        <style>
+                                          <class name="dim-label"/>
+                                        </style>
+                                      </object>
+                                      <packing>
+                                        <property name="expand">False</property>
+                                        <property name="fill">False</property>
+                                        <property name="position">0</property>
+                                      </packing>
+                                    </child>
+                                    <child>
+                                      <object class="GtkScale" id="background-transparent-scale">
+                                        <property name="visible">True</property>
+                                        <property name="can_focus">True</property>
+                                        <property name="adjustment">background-transparent-adjustment</property>
+                                        <property name="draw_value">False</property>
+                                      </object>
+                                      <packing>
+                                        <property name="expand">True</property>
+                                        <property name="fill">True</property>
+                                        <property name="position">1</property>
+                                      </packing>
+                                    </child>
+                                    <child>
+                                      <object class="GtkLabel" id="background-transparent-max-label">
+                                        <property name="visible">True</property>
+                                        <property name="can_focus">False</property>
+                                        <property name="xalign">0.5</property>
+                                        <property name="label" translatable="yes">full</property>
+                                        <style>
+                                          <class name="dim-label"/>
+                                        </style>
+                                      </object>
+                                      <packing>
+                                        <property name="expand">False</property>
+                                        <property name="fill">False</property>
+                                        <property name="position">2</property>
+                                      </packing>
+                                    </child>
+                                  </object>
+                                  <packing>
+                                    <property name="expand">True</property>
+                                    <property name="fill">True</property>
+                                    <property name="position">1</property>
+                                  </packing>
+                                </child>
+                              </object>
+                              <packing>
+                                <property name="expand">True</property>
+                                <property name="fill">True</property>
+                                <property name="position">2</property>
+                              </packing>
+                            </child>
+                            <child>
+                              <object class="GtkBox" id="use-theme-transparency-box">
+                                <property name="visible">True</property>
+                                <property name="can_focus">False</property>
+                                <property name="orientation">horizontal</property>
+                                <property name="spacing">12</property>
+                                <child>
+                                  <object class="GtkCheckButton" id="use-theme-transparency-checkbutton">
+                                    <property name="label" translatable="yes">Use transparency from system theme</property>
+                                    <property name="visible">True</property>
+                                    <property name="can_focus">True</property>
+                                    <property name="receives_default">False</property>
+                                    <property name="use_underline">True</property>
+                                    <property name="xalign">0</property>
+                                    <property name="draw_indicator">True</property>
+                                  </object>
+                                  <packing>
+                                    <property name="expand">False</property>
+                                    <property name="fill">False</property>
+                                    <property name="position">0</property>
+                                  </packing>
+                                </child>
+                              </object>
+                              <packing>
+                                <property name="expand">True</property>
+                                <property name="fill">True</property>
+                                <property name="position">3</property>
+                              </packing>
+                            </child>
                           </object>
                         </child>
                       </object>
Index: b/src/terminal-schemas.h
===================================================================
--- a/src/terminal-schemas.h
+++ b/src/terminal-schemas.h
@@ -69,6 +69,10 @@
 #define TERMINAL_PROFILE_VISIBLE_NAME_KEY               "visible-name"
 #define TERMINAL_PROFILE_WORD_CHAR_EXCEPTIONS_KEY       "word-char-exceptions"
 
+#define TERMINAL_PROFILE_USE_TRANSPARENT_BACKGROUND      "use-transparent-background"
+#define TERMINAL_PROFILE_USE_THEME_TRANSPARENCY          "use-theme-transparency"
+#define TERMINAL_PROFILE_BACKGROUND_TRANSPARENCY_PERCENT "background-transparency-percent"
+
 #define TERMINAL_SETTING_CONFIRM_CLOSE_KEY              "confirm-close"
 #define TERMINAL_SETTING_DEFAULT_SHOW_MENUBAR_KEY       "default-show-menubar"
 #define TERMINAL_SETTING_ENABLE_MENU_BAR_ACCEL_KEY      "menu-accelerator-enabled"
Index: b/src/terminal-screen.c
===================================================================
--- a/src/terminal-screen.c
+++ b/src/terminal-screen.c
@@ -606,6 +606,11 @@
 
   g_type_class_add_private (object_class, sizeof (TerminalScreenPrivate));
 
+  gtk_widget_class_install_style_property (widget_class,
+     g_param_spec_float ("background-darkness", NULL, NULL, -1, 1, -1,
+                         G_PARAM_READABLE | G_PARAM_STATIC_NAME | G_PARAM_STATIC_NICK | G_PARAM_STATIC_BLURB));
+
+
   n_url_regexes = G_N_ELEMENTS (url_regex_patterns);
   precompile_regexes (url_regex_patterns, n_url_regexes, &url_regexes, &url_regex_flavors);
   n_extra_regexes = G_N_ELEMENTS (extra_regex_patterns);
@@ -870,7 +875,10 @@
       prop_name == I_(TERMINAL_PROFILE_HIGHLIGHT_COLORS_SET_KEY) ||
       prop_name == I_(TERMINAL_PROFILE_HIGHLIGHT_BACKGROUND_COLOR_KEY) ||
       prop_name == I_(TERMINAL_PROFILE_HIGHLIGHT_FOREGROUND_COLOR_KEY) ||
-      prop_name == I_(TERMINAL_PROFILE_PALETTE_KEY))
+      prop_name == I_(TERMINAL_PROFILE_PALETTE_KEY) ||
+      prop_name == I_(TERMINAL_PROFILE_USE_TRANSPARENT_BACKGROUND) ||
+      prop_name == I_(TERMINAL_PROFILE_BACKGROUND_TRANSPARENCY_PERCENT) ||
+      prop_name == I_(TERMINAL_PROFILE_USE_THEME_TRANSPARENCY))
     update_color_scheme (screen);
 
   if (!prop_name || prop_name == I_(TERMINAL_PROFILE_AUDIBLE_BELL_KEY))
@@ -941,6 +949,9 @@
   GdkRGBA *highlight_bgp = NULL, *highlight_fgp = NULL;
   GtkStyleContext *context;
   gboolean use_theme_colors;
+  GtkWidget *toplevel;
+  gboolean transparent, theme_transparent;
+  gfloat style_darkness;
 
   context = gtk_widget_get_style_context (widget);
   gtk_style_context_get_color (context, gtk_style_context_get_state (context), &theme_fg);
@@ -992,6 +1003,31 @@
     }
 
   colors = terminal_g_settings_get_rgba_palette (priv->profile, TERMINAL_PROFILE_PALETTE_KEY, &n_colors);
+  theme_transparent = g_settings_get_boolean (profile, TERMINAL_PROFILE_USE_THEME_TRANSPARENCY);
+  transparent = g_settings_get_boolean (profile, TERMINAL_PROFILE_USE_TRANSPARENT_BACKGROUND);
+
+  gtk_widget_style_get (GTK_WIDGET (screen),
+                        "background-darkness", &style_darkness,
+                        NULL);
+
+  if (theme_transparent && style_darkness >= 0)
+    {
+      bg.alpha = style_darkness;
+    }
+  else if (transparent)
+    {
+      gint transparency_percent;
+
+      transparency_percent = g_settings_get_int (profile, TERMINAL_PROFILE_BACKGROUND_TRANSPARENCY_PERCENT);
+      bg.alpha = (100 - transparency_percent) / 100.0;
+    }
+  else
+      bg.alpha = 1.0;
+
+  /* If this gets out of range, let's not crash */
+  if (bg.alpha < 0.0 || bg.alpha > 1.0)
+      bg.alpha = 1.0;
+
   vte_terminal_set_colors (VTE_TERMINAL (screen), &fg, &bg,
                            colors, n_colors);
   vte_terminal_set_color_bold (VTE_TERMINAL (screen), boldp);
@@ -999,6 +1035,10 @@
   vte_terminal_set_color_cursor_foreground (VTE_TERMINAL (screen), cursor_fgp);
   vte_terminal_set_color_highlight (VTE_TERMINAL (screen), highlight_bgp);
   vte_terminal_set_color_highlight_foreground (VTE_TERMINAL (screen), highlight_fgp);
+
+  toplevel = gtk_widget_get_toplevel (GTK_WIDGET (screen));
+  if (toplevel != NULL && gtk_widget_is_toplevel (toplevel))
+    gtk_widget_set_app_paintable (toplevel, transparent);
 }
 
 static void
Index: b/src/terminal-window.c
===================================================================
--- a/src/terminal-window.c
+++ b/src/terminal-window.c
@@ -2303,6 +2303,26 @@
 }
 
 static gboolean
+terminal_window_draw (GtkWidget *widget,
+                      cairo_t   *cr)
+{
+  if (gtk_widget_get_app_paintable (widget))
+    {
+      GtkStyleContext *context;
+      int width;
+      int height;
+
+      context = gtk_widget_get_style_context (widget);
+      width = gtk_widget_get_allocated_width (widget);
+      height = gtk_widget_get_allocated_height (widget);
+      gtk_render_background (context, cr, 0, 0, width, height);
+      gtk_render_frame (context, cr, 0, 0, width, height);
+    }
+
+  return GTK_WIDGET_CLASS (terminal_window_parent_class)->draw (widget, cr);
+}
+
+static gboolean
 terminal_window_state_event (GtkWidget            *widget,
                              GdkEventWindowState  *event)
 {
@@ -2631,6 +2651,8 @@
   TerminalWindowPrivate *priv;
   TerminalApp *app;
   TerminalSettingsList *profiles_list;
+  GdkScreen *screen;
+  GdkVisual *visual;
   GtkActionGroup *action_group;
   GtkAction *action;
   GtkUIManager *manager;
@@ -2645,6 +2667,11 @@
 
   gtk_widget_init_template (GTK_WIDGET (window));
 
+  screen = gtk_widget_get_screen (GTK_WIDGET (window));
+  visual = gdk_screen_get_rgba_visual (screen);
+  if (visual != NULL)
+    gtk_widget_set_visual (GTK_WIDGET (window), visual);
+
   uuid_generate (u);
   uuid_unparse (u, uuidstr);
   priv->uuid = g_strdup (uuidstr);
@@ -2824,6 +2851,7 @@
 
   widget_class->show = terminal_window_show;
   widget_class->realize = terminal_window_realize;
+  widget_class->draw = terminal_window_draw;
   widget_class->window_state_event = terminal_window_state_event;
   widget_class->screen_changed = terminal_window_screen_changed;
   widget_class->style_updated = terminal_window_style_updated;
Index: b/src/migration.c
===================================================================
--- a/src/migration.c
+++ b/src/migration.c
@@ -331,6 +331,9 @@
   migrate_string_list (client, GCONF_GLOBAL_PREFIX, "active_encodings",
                        settings, TERMINAL_SETTING_ENCODINGS_KEY);
 
+  if (verbose)
+      g_printerr("Migrating global\n");
+
   g_object_unref (client);
 
   return TRUE;
@@ -378,6 +381,8 @@
 {
   GSettings *settings;
   char *child_name, *path;
+  gchar *background_type = NULL, *bg_type_key, *bg_darkness_key;
+  gdouble background_darkness;
   gs_free char *name;
   gboolean is_default;
 
@@ -394,6 +399,31 @@
 
   path = gconf_concat_dir_and_key (GCONF_PROFILES_PREFIX, gconf_id);
 
+  bg_type_key = gconf_concat_dir_and_key (path, KEY_BACKGROUND_TYPE);
+  background_type = gconf_client_get_string (client, bg_type_key, NULL);
+  g_free (bg_type_key);
+  if (background_type)
+      g_settings_set_boolean (settings, TERMINAL_PROFILE_USE_TRANSPARENT_BACKGROUND,
+                              g_strcmp0(background_type, "transparent") == 0);
+  g_free (background_type);
+
+  // If we're migrating, set the 'use theme transparency' key to false, this
+  // will be copied from the old profile into transparency.
+  g_settings_set_boolean (settings, TERMINAL_PROFILE_USE_THEME_TRANSPARENCY,
+                          FALSE);
+
+  bg_darkness_key = gconf_concat_dir_and_key (path, KEY_BACKGROUND_DARKNESS);
+  background_darkness = gconf_client_get_float (client,
+                                                bg_darkness_key, NULL);
+
+  g_free (bg_darkness_key);
+  if (background_darkness) {
+      // Need to invert, the key has changed sense
+      g_settings_set_int (settings,
+                          TERMINAL_PROFILE_BACKGROUND_TRANSPARENCY_PERCENT,
+                          1 - (int) (background_darkness * 100));
+  }
+
   migrate_string (client, path, KEY_VISIBLE_NAME,
                   settings, TERMINAL_PROFILE_VISIBLE_NAME_KEY);
 
@@ -481,7 +511,7 @@
   TerminalSettingsList *list;
   GConfClient *client;
   GConfValue *value, *dvalue;
-  GSList *l;
+  GSList *l, *def = NULL;
   const char *default_profile;
 
   client = gconf_client_get_default ();
@@ -491,15 +521,24 @@
       dvalue->type == GCONF_VALUE_STRING)
     default_profile = gconf_value_get_string (dvalue);
   else
-    default_profile = NULL;
+    default_profile = "Default";
 
   list = terminal_profiles_list_new ();
 
   value = gconf_client_get (client, GCONF_GLOBAL_PREFIX "/profile_list", NULL);
+  if (value == NULL) {
+      value = gconf_value_new (GCONF_VALUE_LIST);
+      gconf_value_set_list_type (value, GCONF_VALUE_STRING);
+      def = g_slist_append(def, gconf_value_new_from_string(GCONF_VALUE_STRING,
+                                                            "Default", NULL));
+      gconf_value_set_list_nocopy (value, def);
+  }
   if (value != NULL &&
       value->type == GCONF_VALUE_LIST &&
       gconf_value_get_list_type (value) == GCONF_VALUE_STRING) {
     for (l = gconf_value_get_list (value); l != NULL; l = l->next) {
+      if (verbose)
+          g_printerr ("Migrating %s\n", gconf_value_get_string (l->data));
       migrate_profile (list, client,
                        gconf_value_get_string (l->data),
                        default_profile);
Index: b/src/terminal-app.c
===================================================================
--- a/src/terminal-app.c
+++ b/src/terminal-app.c
@@ -232,7 +232,7 @@
   if (profile == NULL)
     return;
 
-  terminal_profile_edit (profile, NULL, "profile-name-entry");
+  terminal_profile_edit (profile, transient_parent, "profile-name-entry");
 }
 
 void
@@ -669,21 +669,21 @@
                            GtkWindow       *transient_parent,
                            const char      *widget_name)
 {
-  terminal_profile_edit (profile, NULL, widget_name);
+  terminal_profile_edit (profile, transient_parent, widget_name);
 }
 
 void
 terminal_app_edit_preferences (TerminalApp     *app,
                                GtkWindow       *transient_parent)
 {
-  terminal_prefs_show_preferences (NULL, "general");
+  terminal_prefs_show_preferences (transient_parent, "general");
 }
 
 void
 terminal_app_edit_encodings (TerminalApp     *app,
                              GtkWindow       *transient_parent)
 {
-  terminal_prefs_show_preferences (NULL, "encodings");
+  terminal_prefs_show_preferences (transient_parent, "encodings");
 }
 
 /**
Index: b/src/terminal-prefs.c
===================================================================
--- a/src/terminal-prefs.c
+++ b/src/terminal-prefs.c
@@ -300,7 +300,7 @@
   if (selected_profile == NULL)
     return;
 
-  terminal_app_edit_profile (terminal_app_get (), selected_profile, NULL, NULL);
+  terminal_app_edit_profile (terminal_app_get (), selected_profile, GTK_WINDOW (data->parent), NULL);
 }
 
 static GtkTreeView *
@@ -425,7 +425,7 @@
   if (selected_profile == NULL)
     return;
 
-  terminal_app_edit_profile (terminal_app_get (), selected_profile, NULL, NULL);
+  terminal_app_edit_profile (terminal_app_get (), selected_profile, GTK_WINDOW (data->parent), NULL);
 }
 
 static void
